pkg/agent/multicast/mcast_controller_test.go

```diff
@@ -280,3 +341,90 @@ func (c *Controller) initialize(t *testing.T) error {
 	mockMulticastSocket.EXPECT().AllocateVIFs(gomock.Any(), uint16(1)).Times(1).Return([]uint16{1, 2}, nil)
 	return c.Initialize()
 }
+
+func createInterface(name string, ofport uint32) *interfacestore.InterfaceConfig {
+	return &interfacestore.InterfaceConfig{
+		InterfaceName: name,
+		OVSPortConfig: &interfacestore.OVSPortConfig{
+			OFPort: int32(ofport),
+		},
+	}
+}
+
+func createIGMPReportPacketIn(joinedGroups []net.IP, leftGroups []net.IP, version uint8, ofport uint32) []*ofctrl.PacketIn {
+	joinMessages := createIGMPJoinMessage(joinedGroups, version)
+	leftMessages := createIGMPLeaveMessage(leftGroups, version)
+	generatePacket := func(m util.Message) ofctrl.PacketIn {
+		pkt := openflow13.NewPacketIn()
+		matchInport := openflow13.NewInPortField(ofport)
+		pkt.Match.AddField(*matchInport)
+		ipPacket := &protocol.IPv4{
+			Version:  0x4,
+			IHL:      5,
+			Protocol: IGMPProtocolNumber,
+			Length:   20 + m.Len(),
+			Data:     m,
+		}
+		pkt.Data = protocol.Ethernet{
+			HWDst:     pktInDstMAC,
+			HWSrc:     pktInSrcMAC,
+			Ethertype: protocol.IPv4_MSG,
+			Data:      ipPacket,
+		}
+		return ofctrl.PacketIn(*pkt)
+	}
+	pkts := make([]*ofctrl.PacketIn, 0)
+	for _, m := range joinMessages {
+		pkt := generatePacket(m)
+		pkts = append(pkts, &pkt)
+	}
+	for _, m := range leftMessages {
+		pkt := generatePacket(m)
+		pkts = append(pkts, &pkt)
+	}
+	return pkts
+}
+
+func createIGMPLeaveMessage(groups []net.IP, version uint8) []util.Message {
+	pkts := make([]util.Message, 0)
+	switch version {
+	case 1:
+		for i := range groups {
+			pkts = append(pkts, protocol.NewIGMPv2Leave(groups[i]))
```

Comment:
NewIGMPv1Leave?

Reply:
IGMPv1 doesn't support Leave message, the IGMP router finds that the member has left group by no reply responding IGMP query. I would remove the leave case for IGMPv1.